org.eclipse.xtext.java/src/org/eclipse/xtext/java/resource/ClassFileCache.java

```diff
@@ -0,0 +1,156 @@
+ * @since 2.15
+ */
+public class ClassFileCache {
+	public static class ClassFileCacheAdapter extends AdapterImpl {
+		private ClassFileCache element;
+
+		public ClassFileCacheAdapter(ClassFileCache element) {
+			this.element = element;
+		}
+
+		public ClassFileCache get() {
+			return this.element;
+		}
+
+		@Override
+		public boolean isAdapterForType(Object object) {
+			return object == ClassFileCache.class;
+		}
+	}
+
+	private static final Object NULL = new Object();
+
+	private final Map<QualifiedName, Object> cache = new ConcurrentHashMap<QualifiedName, Object>();
+
+	private final Set<Resource> resourcesToCompile = Collections
+			.<Resource>newSetFromMap(new ConcurrentHashMap<Resource, Boolean>());
+
+	private final Map<char[], List<String>> allTopLevelTypes = new ConcurrentHashMap<char[], List<String>>();
+
+	private final Map<char[], Map<String, byte[]>> allClassMaps = new ConcurrentHashMap<char[], Map<String, byte[]>>();
+
+	public boolean containsKey(QualifiedName qualifiedName) {
+		return cache.containsKey(qualifiedName);
+	}
+
+	public IBinaryType get(QualifiedName qualifiedName) {
+		Object result = cache.get(qualifiedName);
+		if (result == ClassFileCache.NULL) {
+			return null;
+		}
+		return (IBinaryType) result;
+	}
+
+	public void put(QualifiedName qualifiedName, IBinaryType answer) {
+		if (answer == null) {
+			cache.put(qualifiedName, ClassFileCache.NULL);
+		} else {
+			cache.put(qualifiedName, answer);
+		}
+	}
+
+	public IBinaryType computeIfAbsent(QualifiedName qualifiedName,
+			Function<? super QualifiedName, ? extends IBinaryType> fun) {
+		Object result = this.cache.computeIfAbsent(qualifiedName, (QualifiedName fqn) -> {
+			IBinaryType binaryType = fun.apply(fqn);
+			if (binaryType != null) {
+				return binaryType;
+			} else {
+				return ClassFileCache.NULL;
+			}
+		});
+		if (result instanceof IBinaryType) {
+			return (IBinaryType) result;
+		}
+		return null;
+	}
+
+	public void clear() {
+		cache.clear();
+		resourcesToCompile.clear();
+		allTopLevelTypes.clear();
+		allClassMaps.clear();
+	}
+
+	public void addResourceToCompile(Resource resource) {
+		resourcesToCompile.add(resource);
+	}
+
+	public Set<Resource> drainResourcesToCompile() {
+		HashSet<Resource> result = new HashSet<Resource>(resourcesToCompile);
+		resourcesToCompile.clear();
+		return result;
+	}
+
+	public boolean popCompileResult(char[] fileName,
+			Procedure2<? super List<String>, ? super Map<String, byte[]>> consumer) {
+		if (allTopLevelTypes.containsKey(fileName)) {
+			consumer.apply(allTopLevelTypes.remove(fileName), allClassMaps.remove(fileName));
+			return true;
+		}
+		return false;
+	}
+
+	public void addCompileResult(char[] fileName, List<String> topLevelTypes, Map<String, byte[]> classMap) {
+		allTopLevelTypes.put(fileName, topLevelTypes);
+		allClassMaps.put(fileName, classMap);
+	}
+
+	public static ClassFileCache findInEmfObject(Notifier emfObject) {
+		for (Adapter adapter : emfObject.eAdapters()) {
```

Comment:
Should use EcoreUtil.getExistingAdapter to benefit from the new EScannableAdapterList